File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/scoutlist/viewholder/StopwatchViewHolder.java
package com.supercilex.robotscouter.ui.scouting.scoutlist.viewholder;

import android.graphics.Color;
import android.os.Build;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.transition.AutoTransition;
import android.support.transition.Transition;
import android.support.transition.TransitionManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.ContextMenu;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;
import com.github.rubensousa.gravitysnaphelper.GravitySnapHelper;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.crash.FirebaseCrash;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.model.Metric;
import com.supercilex.robotscouter.ui.scouting.MetricViewHolderBase;
import com.supercilex.robotscouter.util.AsyncTaskExecutor;
import com.supercilex.robotscouter.util.ui.RecyclerPoolHolder;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class StopwatchViewHolder extends MetricViewHolderBase<Metric<List<Long>>, List<Long>, TextView>
        implements View.OnClickListener {
    private static final Map<Metric.Stopwatch, Timer> TIMERS = new ConcurrentHashMap<>();

    private final Button mToggleStopwatch;
    private final RecyclerView mCycles;

    private Timer mTimer;

    public StopwatchViewHolder(View itemView) {
        super(itemView);
        mToggleStopwatch = itemView.findViewById(R.id.stopwatch);
        mCycles = itemView.findViewById(R.id.list);

        mToggleStopwatch.setOnClickListener(this);

        LinearLayoutManager manager = new LinearLayoutManager(
                itemView.getContext(), LinearLayoutManager.HORIZONTAL, false);
        manager.setInitialPrefetchItemCount(5);
        mCycles.setLayoutManager(manager);
        mCycles.setAdapter(new Adapter());
        new GravitySnapHelper(Gravity.START).attachToRecyclerView(mCycles);

        FragmentManager fragmentManager = ((FragmentActivity) itemView.getContext()).getSupportFragmentManager();
        for (Fragment fragment : fragmentManager.getFragments()) {
            if (fragment instanceof RecyclerPoolHolder) {
                mCycles.setRecycledViewPool(((RecyclerPoolHolder) fragment).getRecyclerPool());
                break;
            }
        }
    }

    @Override
    protected void bind() {
        super.bind();
        setText(R.string.metric_stopwatch_start_title);

        mCycles.setHasFixedSize(false);
        mCycles.getAdapter().notifyDataSetChanged();

        Timer timer = TIMERS.get(getMetric());
        if (timer != null) {
            timer.setHolder(this);
            mTimer = timer;
            mTimer.updateButtonTime();
        }
    }

    @Override
    public void onClick(View v) {
        if (mTimer == null) {
            setText(R.string.metric_stopwatch_stop_title, "0:00");
            mTimer = new Timer(this);
        } else {
            ArrayList<Long> newCycles = new ArrayList<>(getMetric().getValue());
            newCycles.add(mTimer.cancel());
            getMetric().setValue(newCycles);

            RecyclerView.Adapter adapter = mCycles.getAdapter();
            int size = getMetric().getValue().size();

            // Force RV to request layout when adding first item
            mCycles.setHasFixedSize(size > 1);

            if (size == 2) {
                // Add the average metric
                adapter.notifyItemInserted(0);
                adapter.notifyItemInserted(size);
            } else {
                // Account for the average card being there or not
                adapter.notifyItemInserted(size == 1 ? 0 : size);
                adapter.notifyItemChanged(0);
            }
        }
    }

    private void setText(@StringRes int id, Object... formatArgs) {
        mToggleStopwatch.setText(itemView.getResources().getString(id, formatArgs));
    }

    private static final class Timer implements OnSuccessListener<Void>, OnFailureListener {
        private static final int GAME_TIME = 3;
        private static final String COLON = ":";
        private static final String LEADING_ZERO = "0";

        private static Transition sTransition;

        private final long mStartTimeMillis = System.currentTimeMillis();

        private TaskCompletionSource<Void> mTimerTask;
        private boolean mIsRunning;

        private WeakReference<StopwatchViewHolder> mHolder;

        static {
            Transition transition = new AutoTransition();
            transition.excludeTarget(RecyclerView.class, true);
            sTransition = transition;
        }

        public Timer(StopwatchViewHolder holder) {
            mHolder = new WeakReference<>(holder);
            mIsRunning = true;
            TIMERS.put((Metric.Stopwatch) (Object) holder.getMetric(), this);

            updateStyle();

            Tasks.<Void>forResult(null).addOnSuccessListener(AsyncTaskExecutor.INSTANCE, this);
        }

        public static String getFormattedTime(long nanos) {
            long minutes = TimeUnit.MILLISECONDS.toMinutes(nanos);

            String formattedTime =
                    minutes + COLON + (TimeUnit.MILLISECONDS.toSeconds(nanos) - (minutes * 60));
            String[] split = formattedTime.split(COLON);
            if (split[1].length() <= 1) { // NOPMD
                formattedTime = split[0] + COLON + LEADING_ZERO + split[1];
            }

            return formattedTime;
        }

        public void setHolder(StopwatchViewHolder holder) {
            mHolder = new WeakReference<>(holder);
            updateStyle();
        }

        public void updateButtonTime() {
            setText(R.string.metric_stopwatch_stop_title, getFormattedTime(getElapsedTime()));
        }

        /** @return The time since this class was instantiated and then cancelled */
        public long cancel() {
            mIsRunning = false;
            StopwatchViewHolder holder = mHolder.get();
            if (holder != null) holder.mTimer = null;
            for (Map.Entry<Metric.Stopwatch, Timer> entry : TIMERS.entrySet()) {
                if (entry.getValue().equals(this)) TIMERS.remove(entry.getKey());
            }

            mTimerTask.trySetException(new CancellationException());
            updateStyle();
            setText(R.string.metric_stopwatch_start_title);

            return getElapsedTime();
        }

        @Override
        public void onSuccess(Void aVoid) {
            mTimerTask = new TaskCompletionSource<>();
            Task<Void> task = mTimerTask.getTask();
            task.addOnSuccessListener(AsyncTaskExecutor.INSTANCE, this);
            task.addOnFailureListener(this);
            try {
                Thread.sleep(TimeUnit.SECONDS.toMillis(1));

                final long seconds = TimeUnit.MILLISECONDS.toSeconds(getElapsedTime());
                if (!mIsRunning || TimeUnit.SECONDS.toMinutes(seconds) >= GAME_TIME) {
                    mTimerTask.trySetException(
                            new TimeoutException("Cycle time is longer than game time."));
                    return;
                }

                StopwatchViewHolder holder = mHolder.get();
                if (holder != null) {
                    new Handler(holder.itemView.getContext()
                                        .getMainLooper()).post(this::updateButtonTime);
                }
            } catch (InterruptedException e) {
                FirebaseCrash.report(e);
                Crashlytics.logException(e);
                mTimerTask.setException(e);
            }
            mTimerTask.setResult(null);
        }

        @Override
        public void onFailure(@NonNull Exception e) {
            cancel();
        }

        /** @return the time since this class was instantiated in milliseconds */
        private long getElapsedTime() {
            return System.currentTimeMillis() - mStartTimeMillis;
        }

        private void setText(@StringRes int id, Object... formatArgs) {
            StopwatchViewHolder holder = mHolder.get();
            if (holder != null) holder.setText(id, formatArgs);
        }

        private void updateStyle() {
            StopwatchViewHolder holder = mHolder.get();
            if (holder == null) return;

            TransitionManager.beginDelayedTransition((ViewGroup) holder.itemView, sTransition);

            // There's a bug pre-L where changing the view state doesn't update the vector drawable.
            // Because of that, calling View#setActivated(isRunning) doesn't update the background
            // color and we end up with unreadable text.
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;

            Button stopwatch = holder.mToggleStopwatch;
            stopwatch.setTextColor(mIsRunning ? ContextCompat.getColor(
                    stopwatch.getContext(), R.color.colorAccent) : Color.WHITE);
            stopwatch.setActivated(mIsRunning);
            stopwatch.setCompoundDrawablesRelativeWithIntrinsicBounds(
                    mIsRunning ? R.drawable.ic_timer_off_accent_24dp : R.drawable.ic_timer_white_24dp,
                    0,
                    0,
                    0);
        }
    }

    private class Adapter extends RecyclerView.Adapter<DataHolder> {
        // Don't conflict with metric types since the pool is shared
        private static final int DATA_ITEM = 1000;
        private static final int AVERAGE_ITEM = 1001;

        @Override
        public DataHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.scout_stopwatch_cycle_item, parent, false);

            ViewGroup.LayoutParams params = view.getLayoutParams();
            params.width = ViewGroup.LayoutParams.WRAP_CONTENT;
            view.setLayoutParams(params);

            return viewType == DATA_ITEM ? new DataHolder(view) : new AverageHolder(view);
        }

        @Override
        public void onBindViewHolder(DataHolder holder, int position) {
            List<Long> cycles = getMetric().getValue();
            if (holder instanceof AverageHolder) {
                ((AverageHolder) holder).bind(cycles);
            } else {
                if (containsAverageItems()) {
                    int realIndex = position - 1;
                    holder.bind(StopwatchViewHolder.this, cycles.get(realIndex));
                } else {
                    holder.bind(StopwatchViewHolder.this, cycles.get(position));
                }
            }
        }

        @Override
        public int getItemCount() {
            int size = getMetric().getValue().size();
            return containsAverageItems() ? size + 1 : size;
        }

        @Override
        public int getItemViewType(int position) {
            return containsAverageItems() && position == 0 ? AVERAGE_ITEM : DATA_ITEM;
        }

        public boolean containsAverageItems() {
            return getMetric().getValue().size() > 1;
        }
    }

    private static class DataHolder extends RecyclerView.ViewHolder
            implements View.OnCreateContextMenuListener, MenuItem.OnMenuItemClickListener {
        protected TextView mTitle;
        protected TextView mValue;

        private StopwatchViewHolder mHolder;

        public DataHolder(View itemView) {
            super(itemView);
            mTitle = itemView.findViewById(android.R.id.text1);
            mValue = itemView.findViewById(android.R.id.text2);

            itemView.setOnCreateContextMenuListener(this);
        }

        public void bind(StopwatchViewHolder holder, long nanoTime) {
            mHolder = holder;

            mTitle.setText(itemView.getContext().getString(
                    R.string.metric_stopwatch_cycle_title, getRealPosition() + 1));
            mValue.setText(Timer.getFormattedTime(nanoTime));
        }

        @Override
        public void onCreateContextMenu(ContextMenu menu,
                                        View v,
                                        ContextMenu.ContextMenuInfo menuInfo) {
            menu.add(R.string.delete).setOnMenuItemClickListener(this);
        }

        @Override
        public boolean onMenuItemClick(MenuItem item) {
            Metric<List<Long>> metric = mHolder.getMetric();
            boolean hadAverage = hasAverage();

            ArrayList<Long> newCycles = new ArrayList<>(metric.getValue());
            newCycles.remove(getRealPosition());
            metric.setValue(newCycles);

            RecyclerView.Adapter adapter = mHolder.mCycles.getAdapter();
            int size = metric.getValue().size();

            // Force RV to request layout when removing last item
            mHolder.mCycles.setHasFixedSize(size > 0);

            adapter.notifyItemRemoved(getAdapterPosition());
            if (hadAverage && size == 1) {
                // Remove the average card
                adapter.notifyItemRemoved(0);
            } else if (size > 1) {
                adapter.notifyItemChanged(0);
            }

            return true;
        }

        private int getRealPosition() {
            return hasAverage() ? getAdapterPosition() - 1 : getAdapterPosition();
        }

        private boolean hasAverage() {
            return ((Adapter) mHolder.mCycles.getAdapter()).containsAverageItems();
        }
    }

    private static class AverageHolder extends DataHolder {
        public AverageHolder(View itemView) {
            super(itemView);
            mTitle.setText(R.string.metric_stopwatch_cycle_average_title);

            itemView.setOnCreateContextMenuListener(null);
        }

        public void bind(List<Long> cycles) {
            long sum = 0;
            for (Long duration : cycles) sum += duration;
            long nanos = cycles.isEmpty() ? sum : sum / cycles.size();
            mValue.setText(Timer.getFormattedTime(nanos));
        }
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/scoutlist/viewholder/StopwatchViewHolder.kt
package com.supercilex.robotscouter.ui.scouting.scoutlist.viewholder

import android.os.Build
import android.support.annotation.StringRes
import android.support.design.widget.BaseTransientBottomBar
import android.support.design.widget.Snackbar
import android.support.transition.AutoTransition
import android.support.transition.TransitionManager
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.ContextMenu
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import com.github.rubensousa.gravitysnaphelper.GravitySnapHelper
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.RobotScouter
import com.supercilex.robotscouter.data.model.Metric
import com.supercilex.robotscouter.ui.scouting.MetricViewHolderBase
import com.supercilex.robotscouter.util.reportOrCancel
import com.supercilex.robotscouter.util.second
import com.supercilex.robotscouter.util.ui.RecyclerPoolHolder
import com.supercilex.robotscouter.util.ui.setOnLongClickListenerCompat
import kotterknife.bindView
import org.jetbrains.anko.design.longSnackbar
import org.jetbrains.anko.find
import org.jetbrains.anko.runOnUiThread
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Future
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

open class StopwatchViewHolder(
        itemView: View
) : MetricViewHolderBase<Metric.Stopwatch, List<Long>, TextView>(itemView),
        View.OnClickListener, View.OnLongClickListener {
    private val toggleStopwatch: Button by bindView(R.id.stopwatch)
    private val cycles: RecyclerView by bindView(R.id.list)

    private var timer: Timer? = null
    private var undoAddSnackbar: Snackbar? = null

    init {
        toggleStopwatch.setOnClickListener(this)
        toggleStopwatch.setOnLongClickListenerCompat(this)

        cycles.layoutManager = LinearLayoutManager(
                itemView.context,
                LinearLayoutManager.HORIZONTAL,
                false
        ).apply {
            initialPrefetchItemCount = 6
        }
        cycles.adapter = Adapter()
        GravitySnapHelper(Gravity.START).attachToRecyclerView(cycles)

        (itemView.context as FragmentActivity).supportFragmentManager.fragments
                .filterIsInstance<RecyclerPoolHolder>()
                .single()
                .let { cycles.recycledViewPool = it.recyclerPool }
    }

    override fun bind() {
        super.bind()
        cycles.setHasFixedSize(false)
        cycles.adapter.notifyDataSetChanged()

        val timer = TIMERS[metric]
        if (timer == null) {
            setText(R.string.metric_stopwatch_start_title)
            updateStyle(false)
        } else {
            timer.holder = this
            updateStyle(true)
            timer.updateButtonTime()
            this.timer = timer
        }
    }

    override fun onClick(v: View) {
        val currentTimer = timer
        if (currentTimer == null) {
            undoAddSnackbar?.dismiss()
            timer = Timer(this)
        } else {
            metric.value = metric.value.toMutableList().apply {
                add(currentTimer.cancel())
            }

            val size = metric.value.size
            notifyCycleAdded(size, size)

            longSnackbar(itemView, R.string.scout_stopwatch_lap_added_message, R.string.undo) {
                val hadAverage = metric.value.size >= LIST_SIZE_WITH_AVERAGE
                val newCycles = metric.value.toMutableList().apply {
                    removeAt(size - 1)
                }
                metric.value = newCycles

                notifyCycleRemoved(metric.value.size, metric.value.size, hadAverage)
                timer = Timer(this, currentTimer.startTimeMillis)
            }
        }
    }

    override fun onLongClick(v: View): Boolean {
        val currentTimer = timer ?: return false

        currentTimer.cancel()
        undoAddSnackbar = longSnackbar(itemView, R.string.cancelled, R.string.undo) {
            timer = Timer(this, currentTimer.startTimeMillis)
        }.addCallback(object : BaseTransientBottomBar.BaseCallback<Snackbar>() {
            override fun onDismissed(transientBottomBar: Snackbar, event: Int) {
                undoAddSnackbar = null
            }
        })

        return true
    }

    private fun notifyCycleAdded(position: Int, size: Int) {
        // Force RV to request layout when adding first item
        cycles.setHasFixedSize(size >= LIST_SIZE_WITH_AVERAGE)

        val adapter = cycles.adapter
        if (size == LIST_SIZE_WITH_AVERAGE) {
            // Add the average card
            adapter.notifyItemInserted(0)
            adapter.notifyItemInserted(position)
        } else {
            // Account for the average card being there or not. Since we are adding a new lap,
            // there are only two possible states: 1 item or n + 2 items.
            adapter.notifyItemInserted(if (size == 1) 0 else position)
            // Ensure the average card is updated if it's there
            adapter.notifyItemChanged(0)
        }
    }

    private fun notifyCycleRemoved(position: Int, size: Int, hadAverage: Boolean) {
        // Force RV to request layout when removing last item
        cycles.setHasFixedSize(size > 0)

        val adapter = cycles.adapter
        adapter.notifyItemRemoved(if (hadAverage) position + 1 else position)
        if (hadAverage && size == 1) {
            // Remove the average card
            adapter.notifyItemRemoved(0)
        } else if (size >= LIST_SIZE_WITH_AVERAGE) {
            adapter.notifyItemChanged(0) // Ensure the average card is updated
        }
    }

    private fun setText(@StringRes id: Int, vararg formatArgs: Any) {
        toggleStopwatch.text = itemView.resources.getString(id, *formatArgs)
    }

    private fun updateStyle(isRunning: Boolean) {
        // There's a bug pre-L where changing the view state doesn't update the vector drawable.
        // Because of that, calling View#setActivated(isRunning) doesn't update the background
        // color and we end up with unreadable text.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toggleStopwatch.isActivated = isRunning
        }
    }

    private class Timer(
            holder: StopwatchViewHolder,
            val startTimeMillis: Long = System.currentTimeMillis()
    ) : Runnable {
        private val updater: Future<*>

        /**
         * The ID of the metric who originally started the request. We need to validate it since
         * ViewHolders may be recycled across different instances.
         */
        private val metric = holder.metric

        private var _holder: WeakReference<StopwatchViewHolder> = WeakReference(holder)
        var holder: StopwatchViewHolder?
            get() = _holder.get()?.takeIf { it.metric.ref == metric.ref }
            set(holder) {
                _holder = WeakReference<StopwatchViewHolder>(holder)
            }

        /** @return the time since this class was instantiated in milliseconds */
        private val elapsedTime: Long
            get() = System.currentTimeMillis() - startTimeMillis

        init {
            TIMERS[holder.metric] = this
            updater = executor.scheduleWithFixedDelay(this, 1, 1, TimeUnit.SECONDS)
            updateStyle()
            updateButtonTime()
        }

        override fun run() {
            if (TimeUnit.SECONDS.toMinutes(TimeUnit.MILLISECONDS.toSeconds(elapsedTime)) >= GAME_TIME) {
                cancel()
                return
            }

            RobotScouter.runOnUiThread { updateButtonTime() }
        }

        fun updateButtonTime() {
            setText(R.string.metric_stopwatch_stop_title, getFormattedTime(elapsedTime))
        }

        /** @return the time since this class was instantiated and then cancelled */
        fun cancel(): Long {
            holder?.timer = null
            TIMERS.filter { it.value == this }.forEach {
                TIMERS.remove(it.key)
            }

            updater.reportOrCancel(true)
            RobotScouter.runOnUiThread {
                updateStyle()
                setText(R.string.metric_stopwatch_start_title)
            }

            return elapsedTime
        }

        private fun setText(@StringRes id: Int, vararg formatArgs: Any) {
            holder?.setText(id, *formatArgs)
        }

        private fun updateStyle() {
            val holder = holder ?: return
            TransitionManager.beginDelayedTransition(holder.itemView as ViewGroup, transition)
            holder.updateStyle(!updater.isDone)
        }

        private companion object {
            const val GAME_TIME = 3

            val executor = ScheduledThreadPoolExecutor(1)
            val transition = AutoTransition().apply {
                excludeTarget(RecyclerView::class.java, true)
            }
        }
    }

    private inner class Adapter : RecyclerView.Adapter<DataHolder>() {
        val hasAverageItems get() = metric.value.size > 1

        override fun onCreateViewHolder(
                parent: ViewGroup,
                viewType: Int
        ): DataHolder = LayoutInflater.from(parent.context).inflate(
                R.layout.scout_stopwatch_cycle_item,
                parent,
                false
        ).apply {
            layoutParams = layoutParams.apply {
                width = ViewGroup.LayoutParams.WRAP_CONTENT
            }
        }.let {
            if (viewType == DATA_ITEM) DataHolder(it) else AverageHolder(it)
        }

        override fun onBindViewHolder(holder: DataHolder, position: Int) {
            val cycles = metric.value
            when {
                holder is AverageHolder -> holder.bind(cycles)
                hasAverageItems -> {
                    val realIndex = position - 1
                    holder.bind(this@StopwatchViewHolder, cycles[realIndex])
                }
                else -> holder.bind(this@StopwatchViewHolder, cycles[position])
            }
        }

        override fun getItemCount(): Int {
            val size = metric.value.size
            return if (hasAverageItems) size + 1 else size
        }

        override fun getItemViewType(position: Int): Int =
                if (hasAverageItems && position == 0) AVERAGE_ITEM else DATA_ITEM
    }

    private open class DataHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
            View.OnCreateContextMenuListener, MenuItem.OnMenuItemClickListener {
        protected val title: TextView = itemView.find(android.R.id.text1)
        protected val value: TextView = itemView.find(android.R.id.text2)

        /**
         * The outclass's instance. Used indirectly since this ViewHolder may be recycled across
         * different instances.
         */
        private lateinit var holder: StopwatchViewHolder

        private val hasAverage
            get() = (holder.cycles.adapter as Adapter).hasAverageItems
        private val realPosition
            get() = if (hasAverage) adapterPosition - 1 else adapterPosition

        init {
            itemView.setOnCreateContextMenuListener(this)
        }

        fun bind(holder: StopwatchViewHolder, nanoTime: Long) {
            this.holder = holder

            title.text = itemView.context.getString(
                    R.string.metric_stopwatch_cycle_title, realPosition + 1)
            value.text = getFormattedTime(nanoTime)
        }

        override fun onCreateContextMenu(
                menu: ContextMenu,
                v: View,
                menuInfo: ContextMenu.ContextMenuInfo?
        ) {
            menu.add(R.string.delete).setOnMenuItemClickListener(this)
        }

        override fun onMenuItemClick(item: MenuItem): Boolean {
            val metric = holder.metric
            val hadAverage = hasAverage
            val position = realPosition
            val rawPosition = adapterPosition

            val newCycles = metric.value.toMutableList()
            val deletedCycle = newCycles.removeAt(position)
            metric.value = newCycles

            holder.notifyCycleRemoved(position, metric.value.size, hadAverage)

            longSnackbar(itemView, R.string.deleted, R.string.undo) {
                val latestMetric = holder.metric

                latestMetric.value = latestMetric.value.toMutableList().apply {
                    add(position, deletedCycle)
                }

                holder.notifyCycleAdded(rawPosition, latestMetric.value.size)
            }

            return true
        }
    }

    private class AverageHolder(itemView: View) : DataHolder(itemView) {
        init {
            title.setText(R.string.metric_stopwatch_cycle_average_title)
            itemView.setOnCreateContextMenuListener(null)
        }

        fun bind(cycles: List<Long>) {
            value.text = getFormattedTime(cycles.sum() / cycles.size)
        }
    }

    private companion object {
        val TIMERS = ConcurrentHashMap<Metric.Stopwatch, Timer>()

        const val LIST_SIZE_WITH_AVERAGE = 2
        // Don't conflict with metric types since the pool is shared
        const val DATA_ITEM = 1000
        const val AVERAGE_ITEM = 1001

        private const val COLON = ":"
        private const val LEADING_ZERO = "0"

        fun getFormattedTime(nanos: Long): String {
            val minutes = TimeUnit.MILLISECONDS.toMinutes(nanos)
            return (minutes.toString() + COLON +
                    (TimeUnit.MILLISECONDS.toSeconds(nanos) - minutes * 60)).let {
                val split = it.split(COLON.toRegex()).dropLastWhile {
                    it.isEmpty()
                }.toTypedArray()
                if (split.second().length > 1) {
                    it
                } else {
                    split.first() + COLON + LEADING_ZERO + split.second()
                }
            }
        }
    }
}
